quantization/src/encoded_vectors_pq.rs

```diff
@@ -0,0 +1,575 @@
+        let mut result = vec![vec![]; centroids_count];
+
+        // if there are not enough vectors, set centroids as point positions
+        if vector_parameters.count <= centroids_count {
+            for (i, vector_data) in data.into_iter().enumerate() {
+                result[i] = vector_data.to_vec();
+            }
+            for r in result
+                .iter_mut()
+                .take(centroids_count)
+                .skip(vector_parameters.count)
+            {
+                // fill empty centroids just with zeros
+                *r = vec![0.0; vector_parameters.dim];
+            }
+            return Ok(result);
+        }
+
+        // find random subset of data as random non-intersected indexes
+        let permutor = permutation_iterator::Permutor::new(vector_parameters.count as u64);
+        let mut selected_vectors: Vec<usize> =
+            permutor.map(|i| i as usize).take(sample_size).collect();
+        selected_vectors.sort_unstable();
+
+        // find centroids for each chunk
+        for range in vector_division.iter() {
+            // take data subset using indexes from
+            let mut data_subset = Vec::with_capacity(sample_size * range.len());
+            let mut selected_index: usize = 0;
+            for (vector_index, vector_data) in data.clone().enumerate() {
+                if vector_index == selected_vectors[selected_index] {
+                    data_subset.extend_from_slice(&vector_data[range.clone()]);
+                    selected_index += 1;
+                    if selected_index == sample_size {
+                        break;
+                    }
+                }
+            }
+
+            let centroids = kmeans(
+                &data_subset,
+                centroids_count,
+                range.len(),
+                KMEANS_MAX_ITERATIONS,
+                max_kmeans_threads,
+                KMEANS_ACCURACY,
+            )?;
+
+            // push found chunk centroids into result
+            for (centroid_index, centroid_data) in centroids.chunks_exact(range.len()).enumerate() {
+                result[centroid_index].extend_from_slice(centroid_data);
+            }
+        }
+
+        Ok(result)
+    }
+
+    #[cfg(feature = "dump_image")]
+    fn dump_to_image<'a>(
+        data: impl Iterator<Item = &'a [f32]> + Clone,
+        storage: &TStorage,
+        centroids: &[Vec<f32>],
+        vector_division: &[Range<usize>],
+    ) {
+        let (min, max, _count, _dim) =
+            crate::quantile::find_min_max_size_dim_from_iter(data.clone());
+
+        let colors_r: Vec<_> = (0..256).map(|_| rand::random::<u8>()).collect();
+        let colors_g: Vec<_> = (0..256).map(|_| rand::random::<u8>()).collect();
+        let colors_b: Vec<_> = (0..256).map(|_| rand::random::<u8>()).collect();
+        for (range_i, range) in vector_division.iter().enumerate() {
+            if range.len() < 2 {
+                continue;
+            }
+
```

Comment:
I like iterator style. What do you think?

```suggestion
        for (range_i, range) in vector_division
            .iter()
            .enumerate()
            .filter(|(_, range)| range.len() >= 2)
        {
```

Reply:
I like too, done